### PR TITLE
Improve Runtime Description

### PR DESCRIPTION
**Pull request by MarkvanMents**
An overview of how the runtime works and how it is deployed.

### CODE REVIEW

**Comment by MarkvanMents** [comment]
@mijay and @arjenw,
Here is the Runtime document split into four files to be published.
I will publish at the end of the week (17 April) - or earlier if you can both give me your final comments and approvals.

Thanks very much for your help in getting these documents prepared. I have learned a lot about how Mendix apps work and I think it will be very useful to the customers, particularly those new to Mendix, to have an overview of the runtime components.

File: content/refguide/mendix-client.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+## 2 Description
+
+The Mendix Client is a single page application controlled by the client core. This means that all paging is handled by the Mendix Client, rather than being separate pages served to the browser using different URLs. The initial page on which code is bootstrapped is called "shell", and the code consists of the *Core* and *Widgets*.
+
+Because the Mendix Client can run in a browser, everything in the Mendix Client is visible to the end-user. Security is carried out in the Runtime Server, which will only send information to the Mendix Client which the user is allowed to see.
+
+The shell page is created in different ways, depending on the sort of client.
+
+* For a **browser** client, the initial page is provided by a *theme*
+* For a ***mobile app*** the initial page is packaged as an android or iOS app to be downloaded to the mobile device
```
Comment (mijay) [inline content/refguide/mendix-client.md:24]: hybrid app

File: content/refguide/mendix-client.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+This description of the Mendix Client is based on using the Runtime Server of an app running in the cloud. You can also run Mendix locally for testing, but this is conceptually the same.
+
+## 2 Description
+
+The Mendix Client is a single page application controlled by the client core. This means that all paging is handled by the Mendix Client, rather than being separate pages served to the browser using different URLs. The initial page on which code is bootstrapped is called "shell", and the code consists of the *Core* and *Widgets*.
+
+Because the Mendix Client can run in a browser, everything in the Mendix Client is visible to the end-user. Security is carried out in the Runtime Server, which will only send information to the Mendix Client which the user is allowed to see.
+
+The shell page is created in different ways, depending on the sort of client.
```
Comment (mijay) [inline content/refguide/mendix-client.md:21]: There is no mention of native in this document, while now native is quite important

File: content/refguide/mendix-client.md [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+* For a **browser** client, the initial page is provided by a *theme*
+* For a ***mobile app*** the initial page is packaged as an android or iOS app to be downloaded to the mobile device
+
+Below is a chart showing the components of the Mendix Client. Each of the components is described below the chart.
+
+![The makeup of the Mendix Client](attachments/runtime/mendix-client.png)
```
Comment (mijay) [inline content/refguide/mendix-client.md:28]: - I think it is better to rename `Online Data` to something like `Data Api`
- it is `Data Api` that interacts with `Offline Storage`, not `Object Cache`
- Alternatively, introduce `Offline Data` next to `Online Data` and change arrows accordingly (it is the only one to interact with `Offline Storage` and the only one to interact with `Sync` component)

File: content/refguide/mendix-client.md [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+These are the fundamental building blocks of the Mendix Client. All the actions which the client takes are controlled by widgets. They are responsible for what is displayed on pages, and how user input is managed. There is a more detailed description of widgets in [Widgets](#widgets), below.
+
+### 2.3 Javascript Actions
+
+This runs custom JavaScript, added by the app developer, which is held as JavaScript actions in the client config.
+
+### 2.4 UI Layer
+
+The UI layer performs navigation, resource loading, and platform integration. accepts interactions from the end-user (for example, data input or button presses) and passes them to the client core. It is responsible for building the page which is presented to the end-user in response to the actions of the Mendix Client, using the correct language and other locale settings.
```
Comment (mijay) [inline content/refguide/mendix-client.md:49]: `accepts interactions from the end-user and passes them to the client core` is a job for widgets, so imho better omit them here.

File: content/refguide/mendix-client.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+These are functions of the device on which the Mendix Client is running. In most cases this will be a function of a mobile device such as the camera or GPS location, but it can also include making calls to Mendix Native APIs or browser functions such as accessing an image file.
+
+### 2.8 Client Config
+
+This is the static data which is needed by the Mendix Client. For a browser-based client, this data is held online, with the Runtime Server. For native apps, this is held locally in a native bundle.
+
+These include the shell page needed to start the Mendix Client, Cascading Style Sheets (css files) which define the app’s theme, and JavaScript files which define client-side logic.
+
+### 2.9 CodePush
```
Comment (mijay) [inline content/refguide/mendix-client.md:69]: I think CodePush can be removed from the picture and from here. We might stop using it in a not so distant future, so its not worth the publicity.

File: content/refguide/mendix-client.md [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
+This process allows the Mendix app to distribute updates to clients which are running React Native apps.
+
+### 2.10 Online Data
+
+This manages the objects which are maintained in the Mendix Client (non-persistable, new, and changed objects) and ensures that they are passed to the Runtime Server when they are needed as part of a request.
```
Comment (mijay) [inline content/refguide/mendix-client.md:75]: `manages the objects which are maintained in the Mendix Client` is part of Object Cache responisibility

File: content/refguide/mendix-client.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+This holds objects which are being used by the Mendix Client in memory – for example a list of objects returned by the Runtime Server to be displayed on a page. State handling will perform garbage collection to ensure that memory is released when it is no longer needed.
+
+### 2.12 Offline Storage
+
+This is permanent storage, usually on a mobile device, where data can be stored for apps which are running in offline mode. It differs from the temporary object storage in that data here is not lost at the end of a session, but is kept until it can be synced to the Runtime Server.
+
+### 2.13 State/Sync/Session
+
+This manages requests to the Runtime Server. Note that some actions in the Mendix Client will not require access to the Runtime Server. For example, if the Object Manager already has access to the required data in the temporary object storage, or if the app is written as “offline first”.
```
Comment (mijay) [inline content/refguide/mendix-client.md:87]: Object Manager -> Object Cache?

File: content/refguide/mendix-client.md [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+
+For more information about the communication between the Mendix Client and the Runtime Server, see [SIG–Mendix Performance Subjects](sig-mendix-performance-subjects-explanation).
+
+#### 2.13.1 State Handling
+
+This holds the current state of the app and communicates it to the Runtime Server. As the state is held in the Mendix Client, the Runtime Server can be stateless. This ensures that it is easier to scale your app horizontally by adding more instances as any instance can handle any request.
```
Comment (mijay) [inline content/refguide/mendix-client.md:93]: This is also a responsibility of object cache

File: content/refguide/mendix-client.md [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+The state includes the following
+
+* which page is being displayed
```
Comment (mijay) [inline content/refguide/mendix-client.md:96]: no

File: content/refguide/mendix-client.md [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+
+For more detailed information about state, see this blog: [https://www.mendix.com/blog/the-art-of-state-part-1-introduction-to-the-client-state/](https://www.mendix.com/blog/the-art-of-state-part-1-introduction-to-the-client-state/). This also includes a worked example where you can see, and duplicate for yourself, how state is passed to the Runtime Server.
+
+State handling is also responsible for garbage collection. If you want to know more about this aspect, see this blog: [https://www.mendix.com/blog/the-art-of-state-part-2-garbage-collection/](https://www.mendix.com/blog/the-art-of-state-part-2-garbage-collection/).
+
+#### 2.13.2 Synchronization
+
+Where an app is “Offline First”, data created and changed in the app is stored locally until it is synchronized with the Runtime Server. This job is carried out by the synchronization process. This synchronizes the offline storage and object cache with the Runtime Server. For more information on offline first apps and synchronization, see [Offline-First](offline-first).
```
Comment (mijay) [inline content/refguide/mendix-client.md:110]: I think we call it just "offline" elsewhere in documentation, but i might be wrong

File: content/refguide/mendix-client.md [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+The two types of widget are:
+
+* Pluggable widgets
+* Custom widgets
+
+These are described in the sections below.
+
+### 3.1 Pluggable Widgets
+
+Pluggable widgets are reusable UI components written in React or React Native. This is a more powerful language than Dojo, which is used for Custom widgets, and is the recommended way of writing widgets. Pluggable widgets run on the native core framework using the pluggable widgets API which enables different pluggable widgets to communicate. They have limited access to web core widgets, but full access to the native core.
```
Comment (mijay) [inline content/refguide/mendix-client.md:136]: I don't understand sentences 2 and 3.

File: content/refguide/mendix-client.md [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+
+Pluggable widgets have many advantages over Custom widgets. For example, you can use conditional visibility and editability, and can place a Pluggable widget inside another Pluggable widget.
+
+For more information, see [Pluggable Widgets API](/apidocs-mxsdk/apidocs/pluggable-widgets).
+
+### 3.2 Custom Widgets
+
+Custom widgets are written using Dojo and run on the web core framework using a previous version of the client API. Custom widgets can have access to some of the web core widgets through the API. However, they cannot use pluggable widgets or native core widgets which means that they cannot be used in native apps, or in offline hybrid apps. Custom widgets should only be used if you cannot create the functionality in a Pluggable widget.
```
Comment (mijay) [inline content/refguide/mendix-client.md:144]: Also don't understand what do you mean with sentence 2. Custom widgets can be used on offline hybrid.

File: content/refguide/mendix-client.md [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+
+1. The end-user enters the URL of the app in the browser.
+
+2. The browser loads the HTML web page ("shell").
+
+3. The web page loads and starts the Mendix Client, together with the core widgets.
+
+4. The Mendix Client loads any custom widgets.
+
+5. The Mendix Client contacts the Runtime Server and authenticates the end-user.
```
Comment (mijay) [inline content/refguide/mendix-client.md:171]: This applies for anonymous end-users as well.

File: content/refguide/mendix-client.md [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+
+The flow when launching a native app is different from launching in a browser. More information is stored locally as part of the app, and a native app can even be designed to run “offline first”, which means that it can still be run without any connection to the Runtime Server.
+
+The flow described here is for production apps. During development, the flow is not the same. This enables you to do faster deployments and online debugging.
+
+1. The end-user opens the app on their device. This is a project specific shell app, which runs natively on iOS or Android. It is released to the app store appropriate for the device. If a new version of the app is downloaded to the device, the app will behave as if the end-user has opened it, even if it was already open on their device.
+
+2. The shell app loads a native bundle. This is the equivalent of the Mendix Client resources used by the Mendix Client running in a browser. It contains, for example, the Mendix Client code and page definitions. However, it is held locally on the device rather than centrally with the Runtime Server.
+
+3. The Mendix Client contacts the Runtime Server and authenticates the end-user.
```
Comment (mijay) [inline content/refguide/mendix-client.md:211]: If an end-user hasn't been authenticated (this applies to anonymous users as well) yet, the Mendix Client contacts the Runtime Server and authenticates the end-user. Authentication token is stored locally on the device, so this step isn't repeated unless end-users logs in or out in the app.

File: content/refguide/mendix-client.md [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+
+4. The Mendix Client gets any additional configuration required from the Runtime Server.
```
Comment (mijay) [inline content/refguide/mendix-client.md:213]: This configuration is stored locally on the device, so this step isn't repeated unless the end-user logs out from his/her account.

File: content/refguide/mendix-client.md [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+
+5. The Mendix Client performs a synchronization with the Runtime Server.
```
Comment (mijay) [inline content/refguide/mendix-client.md:215]: If this is a first start of the app, or the first start of the app after an update, the Mendix Client performs a synchronization with the Runtime Server.

File: content/refguide/mendix-client.md [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+
+6. The Mendix Client checks the resources stored in Visual Studio App Center for updates to the native bundle. This enables the app to keep up-to-date without needing to download new versions of the app from the app store.
+
+    *The Mendix Client is now ready to start interacting with the end-user and will repeat the following steps for as long as the continues to run.*
+
+7. The Mendix Client prepares a page using either offline or online data.
```
Comment (mijay) [inline content/refguide/mendix-client.md:221]: only offline data

File: content/refguide/mendix-client.md (hunk 3 above)
Comment (MarkvanMents) [reply content/refguide/mendix-client.md:29]: Done. Thanks for the suggestion

File: content/refguide/mendix-client.md (hunk 8 above)
Comment (MarkvanMents) [reply content/refguide/mendix-client.md:92]: Hi @mijay - I've tried to be as accurate as possible about the object cache and state. I realise that internally there is a big distinction between which component does what, however I am trying to give end-users a general idea of what the client is doing and want to make this as clear as I can.
Can you have another quick read through the document and see if I have answered all your concerns?
We can always update this later if there is something which needs to be clarified - so long as nothing is completely wrong (at least from what the customer sees).

File: content/refguide/mendix-client.md [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+
+This runs client-side logic which is defined in the nanoflows in the model.
+
+### 2.7 Platform APIs
+
+These are functions of the device on which the Mendix Client is running. In most cases this will be a function of a mobile device such as the camera or GPS location, but it can also include making calls to Mendix Native APIs or browser functions such as accessing an image file.
+
+### 2.8 Client Config
+
+This is the static data which is needed by the Mendix Client. For a browser-based client, this data is held online, with the Runtime Server. For native apps, this is held locally in a native bundle.
```
Comment (mijay) [inline content/refguide/mendix-client.md:66]: this is held locally on a device.

File: content/refguide/mendix-client.md [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+
+These include the shell page needed to start the Mendix Client, Cascading Style Sheets (css files) which define the app’s theme, and JavaScript files which define client-side logic.
+
+### 2.9 Data API
+
+This allows the Mendix Client to communicate with Offline Storage.
```
Comment (mijay) [inline content/refguide/mendix-client.md:72]: to communicate with Offline Storage or Mendix Runtime. Or instead something like: "This allows the Mendix Client to execute data fetching and manipulation."

File: content/refguide/mendix-client.md (hunk 8 above)
Comment (mijay) [reply content/refguide/mendix-client.md:92]: I would then just remove `holds the current state of the app`: "This communicates the current state of the app to the Runtime Server" maybe with "current state of the app (hold in Object Cache)".

File: content/refguide/mendix-client.md [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+* widgets – both native and web core widgets which are used by this app
+* page definitions – xml page definitions which tell the Mendix Client what the pages for this app look like
+
+### 4.2 Launching Native Mendix Client
+
+The flow when launching a native app is different from launching in a browser. More information is stored locally as part of the app, and a native app can even be designed to run “offline-first”, which means that it can still be run without any connection to the Runtime Server.
+
+The flow described here is for production apps. During development, the flow is not the same. This enables you to do faster deployments and online debugging.
+
+1. The end-user opens the app on their device. This is a project specific shell app, which runs natively on iOS or Android. It is released to the app store appropriate for the device. If a new version of the app is downloaded to the device, the app will behave as if the end-user has opened it, even if it was already open on their device.
```
Comment (mijay) [inline content/refguide/mendix-client.md:207]: Did you mean `as if the end-user has opened it _for the first time_`?

File: content/refguide/mendix-client.md (hunk 19 above)
Comment (MarkvanMents) [reply content/refguide/mendix-client.md:72]: Combined both ideas together. Thanks.

File: content/refguide/mendix-client.md (hunk 8 above)
Comment (MarkvanMents) [reply content/refguide/mendix-client.md:92]: Great - will use that suggestion.

**Pull request merged**